FrEIA/modules/orthogonal.py

```diff
@@ -82,49 +139,32 @@ def __init__(self, dims_in, dims_c=[], n_reflections=1, fixed=False):
                 # init close to identity
                 init = torch.eye(self.width, self.n_reflections)
                 init += torch.randn_like(init) * 0.1
-            Vs = torch.unbind(init, dim=-1)
-            self.Vs = [nn.Parameter(V) for V in Vs]
-            for i, V in enumerate(self.Vs):
-                V.requires_grad = not self.fixed
-                self.register_parameter(f'V_{i}', V)
+            Vs = init.transpose(-1, -2)
+            self.Vs = nn.Parameter(Vs)
+
+            Vs.requires_grad = not self.fixed
+            self.register_parameter(f'Vs', self.Vs)
 
         if self.fixed:
-            I = torch.eye(self.width)
-            self.W = I - 2 * torch.ger(self.Vs[0], self.Vs[0]) / torch.dot(self.Vs[0], self.Vs[0])
-            for i in range(1, self.n_reflections):
-                self.W = self.W.mm(I - 2 * torch.ger(self.Vs[i], self.Vs[i]) / torch.dot(self.Vs[i], self.Vs[i]))
+            self.W = _fast_h(self.Vs)
             self.W = nn.Parameter(self.W, requires_grad=False)
             self.register_parameter('weight', self.W)
 
     def forward(self, x, c=[], rev=False):
 
         if self.conditional:
-            Vs = torch.unbind(c[0].reshape(-1, self.width, self.n_reflections), dim=-1)
-
-            xW = x[0]
-            for i in range(self.n_reflections):
-                if not rev:
-                    V = Vs[i]
-                else:
-                    V = Vs[-i - 1]
-                VVt = torch.matmul(V.unsqueeze(-1), V.unsqueeze(-2))
-                VtV = torch.matmul(V.unsqueeze(-2), V.unsqueeze(-1)).squeeze(-1)
-                xW = xW - torch.matmul(xW.unsqueeze(-2), VVt).squeeze() * (2/VtV)
-            return [xW]
-
+            Vs = c[0].reshape(-1, self.width, self.n_reflections).transpose(-1, -2)
```

Comment:
I don't know if I got this right, does ```self.conditional=True``` mean that the ```v```-vectors are explicitly handed to the function? Is it then correct to calculate ```W``` that way and transpose it in the case of  ```rev=True```

Reply:
Yes, I belive with conditional, the module should be invertible if the condition is kept fix.
So this solution looks right!